src/puppetlabs/puppetdb/mq_listener.clj

```diff
@@ -167,154 +158,88 @@
-    (discard msg nil)))
-
-(defn handle-parse-error
-  [msg e discard]
-  (log/errorf e "Fatal error parsing command: %s" msg)
-  (discard msg e))
-
-(defn handle-command-failure
-  "Dump the error encountered during command-handling to the log and discard
-  the message."
-  [{:keys [command annotations] :as msg} e discard]
-  (let [attempt (count (:attempts annotations))
-        id      (:id annotations)
-        msg     (annotate-with-attempt msg e)]
-    (log/errorf e "[%s] [%s] Fatal error on attempt %d" id command attempt)
-    (discard msg e)))
-
-;; The number of times a message can be retried before we discard it
-(def maximum-allowable-retries 16)
-
-(defn handle-command-retry
-  "Dump the error encountered to the log, and re-publish the message,
-  with an incremented retry counter, after a delay.
-
-  The error is logged to DEBUG level during the first `M/4` retries,
-  and ERROR thereafter, where `M` is the maximum number of allowable
-  retries.
-
-  The retry delay is based on the following exponential backoff
-  algorithm:
-
-    2^(n-1) + random(2^n)
-
-  `n` is the number of the current attempt, and `random` is a random
-  number between 0 and the argument."
-  [{:keys [command version annotations] :as msg} e publish-fn]
-  (mark! (cmd-metric command version :retried))
-  (let [attempt (count (:attempts annotations))
-        id      (:id annotations)
-        msg     (annotate-with-attempt msg e)
-        n       (inc attempt)
-        delay   (+ (Math/pow 2 (dec n))
-                   (rand-int (Math/pow 2 n)))
-        error-msg (format "[%s] [%s] Retrying after attempt %d, due to: %s"
-                          id command attempt e)]
-    (if (> n (/ maximum-allowable-retries 4))
-      (log/error e error-msg)
-      (log/debug e error-msg))
-     (time! (global-metric :retry-persistence-time)
-            (publish-fn (json/generate-string msg) (mq/delay-property delay :seconds)))))
-
-(defn wrap-message-handler-middleware
-  [send-delayed-msg-fn discarded-dir message-fn]
-  (let [discard        #(dlo/store-failed-message %1 %2 discarded-dir)
-        on-discard     #(handle-command-discard % discard)
-        on-parse-error #(handle-parse-error %1 %2 discard)
-        on-fatal       #(handle-command-failure %1 %2 discard)
-        on-retry       #(handle-command-retry %1 %2 send-delayed-msg-fn)]
-    (-> message-fn
-        (wrap-with-discard on-discard maximum-allowable-retries)
-        (wrap-with-exception-handling on-retry on-fatal)
-        (wrap-with-command-parser on-parse-error)
-        (wrap-with-meter (global-metric :seen))
-        (wrap-with-thread-name "command-proc"))))
+          (let [ex (:throwable &throw-context)]
+            (log/errorf ex "[%s] [%s] Fatal error on attempt %d" id command retries)
+            (-> parsed-result
+                (annotate-with-attempt ex)
+                (discard-message-fn ex))))
+        
+        (catch Throwable exception
```

Comment:
Figured out why it was Throwable. There are assertions in the command parsing code. If something like payload is missing, it will throw a `AssertionError` which is not a subclass of `Exception`. I added a specific clause to the parse-command try/catch for that error so that we can continue to handle it like we have before.